homeassistant/components/icloud/sensor.py

```diff
@@ -0,0 +1,96 @@
+"""Support for iCloud sensors."""
+import logging
+from typing import Dict
+
+from homeassistant.config_entries import ConfigEntry
+from homeassistant.const import CONF_USERNAME, DEVICE_CLASS_BATTERY
+from homeassistant.helpers.entity import Entity
+from homeassistant.helpers.icon import icon_for_battery_level
+from homeassistant.helpers.typing import HomeAssistantType
+
+from . import IcloudDevice
+from .const import DOMAIN
+
+_LOGGER = logging.getLogger(__name__)
+
+
+async def async_setup_entry(
+    hass: HomeAssistantType, entry: ConfigEntry, async_add_entities
+) -> None:
+    """Set up iCloud devices sensors based on a config entry."""
+    username = entry.data[CONF_USERNAME]
+    icloud = hass.data[DOMAIN][username]
+
+    devices = []
+    for device_name, icloud_device in icloud.devices.items():
+        if icloud_device.battery_level is not None:
+            _LOGGER.debug("Adding sensors from iCloud device=%s", device_name)
+            devices.append(
+                IcloudDeviceBatterySensor(hass, icloud.username, icloud_device)
+            )
+
+    async_add_entities(devices, True)
+
+
+class IcloudDeviceBatterySensor(Entity):
+    """Representation of a iCloud device battery sensor."""
+
+    def __init__(self, hass: HomeAssistantType, username: str, device: IcloudDevice):
+        """Initialize the battery sensor."""
+        self._hass = hass
+        self._username = username
+        self._device = device
+
+    def update(self):
```

Comment:
If this is just looking up existing data, change the method to be
```suggestion
    async def async_update(self):
```